gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTable.java

```diff
@@ -46,36 +51,89 @@
 public class IcebergTable {
   private final TableOperations tableOps;
 
+  /** @return metadata info limited to the most recent (current) snapshot */
   public IcebergSnapshotInfo getCurrentSnapshotInfo() throws IOException {
     TableMetadata current = tableOps.current();
-    Snapshot snapshot = current.currentSnapshot();
+    return createSnapshotInfo(current.currentSnapshot(), Optional.of(current.metadataFileLocation()));
+  }
+
+  /** @return metadata info for all known snapshots, ordered historically, with *most recent last* */
+  public Iterator<IcebergSnapshotInfo> getAllSnapshotInfosIterator() {
+    TableMetadata current = tableOps.current();
+    long currentSnapshotId = current.currentSnapshot().snapshotId();
+    List<Snapshot> snapshots = current.snapshots();
+    return Iterators.transform(snapshots.iterator(), snapshot -> {
+      try {
+        return IcebergTable.this.createSnapshotInfo(
+            snapshot,
+            currentSnapshotId == snapshot.snapshotId() ? Optional.of(current.metadataFileLocation()) : Optional.empty()
+        );
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    });
+  }
+
+  /**
+   * @return metadata info for all known snapshots, but incrementally, so content overlap between snapshots appears
+   * only within the first as they're ordered historically, with *most recent last*
+   */
+  public Iterator<IcebergSnapshotInfo> getIncrementalSnapshotInfosIterator() {
```

Comment:
presently (this PR too), we only support copying the entire iceberg.  I am already working on the next iteration with is to calculate the delta between the src and dest.  (for integration testing that, we'll bootstrap "destination" with a replicated copy by using this code herein.)

I absolutely agree we do not want to eagerly load the iceberg's entire metadata.  this interface is an `Iterator` for that reason.  I will `.reverse()` the order, to enable walking backwards in history.  the delta comparison will look on the destination for each snapshot's manifest list file.  when that's found we'll infer that replication up to that point was achieved.

so yes, that is a plan I'm in-progress with carrying out.  this commit is just to give us working code against an MVP that does only entire copy (non-incremental, delta copy).

Reply:
It makes sense to me if we are doing the MVP at this point. But if we just want to copy the whole table, why not using a set to store and eliminate duplication directly instead of this complicated logic? As for remaining the order, I believe most of the overhead I'm concerning is the data file, that even just copy current snapshot, you will check all the old data files, comparing to that, metadata files does not seem to be a big deal here?